Overhauled handling of local extensions in Gradle DevTools

vemilyus opened the pull request:
This completely changes how local extensions are handled by the Gradle plugins (`io.quarkus` and `io.quarkus.extension`) by now also resolving deployment modules in included builds and properly adding them and their artifacts as dependencies to the respective Quarkus code generation tasks and configurations. This makes resolution also more resilient by largely depending on Gradle native constructs (`Project`, `ProjectDependency`, etc.) instead of `String` values or artifact coordinates.

The main enhancement is that extensions and deployment modules are now resolved across project boundaries (esp. in composite builds) and more parts of the module hierarchy are searched to find a match. Included build deployment modules are now more reliably detected, even if the extension build isn't explicitly included, and deployment modules that are siblings to the extension module are also more reliably found.

In addition it no longer depends on an extension descriptor file (`META-INF/quarkus-extension.properties`) to exist to determine the deployment module, conditional dependencies and dependency conditions. These are now read directly from the `QuarkusExtensionConfiguration` extension of any local (or included) extension module. Due to Gradle using different ClassLoaders for included builds this had to be done using Reflection. However, if such a properties file is provided by the developer it has precedence over any configuration done using the plugin.

To achieve this deeper integration with Gradle a bit more internal classes were used than I would normally be comfortable with, but they are pretty essential to the functioning of Gradle and Composite Builds so I'm fairly confident that they won't break in the foreseeable future (please give feedback on this).

I also enhanced `QuarkusGradleWrapperTestBase` to offer better usability for local development and testing by starting the test Gradle instance in debug mode if the test is run locally and with a debugger attached. It will then itself wait for a debugger to attach at the default port (`5005`) before doing what it's supposed to do. Also the output of the started Gradle process is printed immediately for better observability and the build is stopped if Quarkus DevMode failed to start three times in a row.

Fixes: #29234

Oh, and merry christmas! :christmas_tree: 

P.S.: Kudos to the people who wrote the Gradle test cases! I didn't have to touch most of them and they were really useful in sorting out most edge cases and bugs.

gsmet commented:
/cc @snazy @glefloch @aloubyansky

snazy commented:
Can take a look a look at this.

vemilyus commented:
Hi, not to be pushy, has there been any progress on the review? I'd really like some feedback on the changes and what needs to be changed/fixed to get this merged. Thanks :)

snazy commented on devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/tooling/ToolingUtils.java:
```diff
@@ -38,21 +43,100 @@ public static boolean isEnforcedPlatform(ModuleDependency module) {
-        final String name = projectComponentIdentifier.getBuild().getName();
+    public static IncludedBuild includedBuild(final Project project, final String buildName) {
+        Gradle currentGradle = project.getRootProject().getGradle();
+        while (null != currentGradle) {
```
When can this happen?

vemilyus replied:
Calling `currentGradle.getParent()` at the bottom of the loop can return `null`, that's our exit condition for walking up the project tree.

snazy commented on devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/tooling/ToolingUtils.java:
```diff
@@ -38,21 +43,100 @@ public static boolean isEnforcedPlatform(ModuleDependency module) {
+        for (Project subproject : project.getRootProject().getSubprojects()) {
+            if (subproject.getGroup().equals(artifactCoords.getGroupId()) &&
+                    subproject.getName().equals(artifactCoords.getArtifactId()) &&
+                    (artifactCoords.getVersion() == null || subproject.getVersion().equals(artifactCoords.getVersion()))) {
```
Nit: Object.equals ?

vemilyus replied:
Gradle projects can all have separate versions, so I want to only explicitly use the version for matching the projects when it is provided in the artifact coordinates. If there is no specified version it should just use the local project as a dependency, if there is a version it should only use the local project if the version matches. Otherwise it would refer to a remote artifact (e.g. Maven dependency)

snazy replied:
NVM - I had `Object.equals(artifactCoords.getVersion(), subproject.getVersion())` in mind - but that doesn't cover the `null` special case

vemilyus committed:
Overhauled handling of local extensions in Gradle Plugins

This completely changes how local extensions are handled by the Gradle plugins (`io.quarkus` and `io.quarkus.extension`) by now also resolving deployment modules in included builds and properly adding them and their artifacts as dependencies to the respective Quarkus code generation tasks and configurations.

In addition it no longer depends on an extension descriptor to exist to determine the deployment module, conditional dependencies and dependency conditions. These are now read directly from the `QuarkusExtensionConfiguration` extension of any local extension module. Due to Gradle using different ClassLoaders for included builds this is done using Reflection.

To achieve this deeper integration with Gradle a bit more internal classes have been used than I would normally be comfortable with, but they are pretty essential to the functioning of Gradle and Composite Builds so I'm fairly confident that they won't break any time soon.

snazy commented on devtools/gradle/gradle-model/src/main/java/io/quarkus/gradle/extension/ConfigurationUtils.java:
```diff
@@ -0,0 +1,50 @@
+// `QuarkusExtensionConfiguration` extension is provided by a different class loader
+// which prevents us from creating some interface or casting to it directly.
+public class ConfigurationUtils {
+    private static final ConcurrentHashMap<String, Method> methodCache = new ConcurrentHashMap<>();
```
Not sure whether a `static final` cache is good, because it would accumulate `Method` and with those `Class` and `ClassLoader` references. I suspect this would also not respect changes to the extensions and re-use old ones.

snazy replied:
Yea - or if needed, cache it for a build, but that's quite tricky w/ Gradle

vemilyus replied:
I removed the cache ;)

snazy reviewed (commented):
Not a full review from my side, because I'm not familiar with the extensions plugin. The changes to the application-plugin look good. Not sure about the `static final` map in `ConfigurationUtils` wrt to multiple Gradle build invocations.

aloubyansky commented on devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/QuarkusPlugin.java:
```diff
@@ -555,13 +559,40 @@ protected void visitProjectDependencies(Project project, Project dep, Set<String
+                        if (depProject.getState().getExecuted()) {
+                            visitPossibleExtensionProject(project, depProject, visited);
+                        } else {
+                            depProject.afterEvaluate(p -> visitPossibleExtensionProject(project, p, visited));
```
Isn't this change limiting dependency processing to local extension subprojects compared to the current impl?

vemilyus replied:
This is called as part of the process of determining a local module's build dependencies, so that they also get built before building the module.

So if I have a module using the application plugin and want to use a local extension, I want to make sure that the local extension and its local module dependencies are also built before building the application.

This doesn't affect how external extensions (e.g. from Maven) are used in any way.

aloubyansky replied:
Right, my question was about visiting non-extension local dependencies.

aloubyansky replied:
It looks like `visitProjectDep` is currently called for any local dependency, while with this change it will be called only for local extension dependencies?

aloubyansky replied:
Yeah, that makes sense now, thanks.

aloubyansky commented on integration-tests/gradle/src/main/resources/basic-composite-build-extension-project/extensions/example-extension/deployment/src/main/java/org/acme/example/extension/deployment/QuarkusExampleProcessor.java:
```diff
@@ -34,13 +31,4 @@ void addLibABean(BuildProducer<AdditionalBeanBuildItem> additionalBeans) {
-		return new EnabledBuildItem(exampleRecorder.create().getValue());
-
-	}
-
```
@tomas1885 would you like to review these changes?

tomas1885 replied:
I think the recorder was there just to make sure that it's that nothing is broken between the runtime and deployment modules. It might be redundant so maybe it's ok to remove it, as long as we're sure the current test cases covers this.

aloubyansky reviewed (approved):
Thanks a lot @vemilyus
